cmasher/utils.py

```diff
@@ -233,6 +239,115 @@ def _get_cmap_perceptual_rank(
 
 
 # %% FUNCTIONS
+# This function combines multiple colormaps at given nodes
+def combine_cmaps(
+    cmaps: list[Union[Colormap, str]],
+    *,
+    nodes: Optional[list[float]] = None,
+    n_rgb_levels: int = 256,
+    combined_cmap_name: str = "combined_cmap",
+) -> LinearSegmentedColormap:
+    """Create a composite matplotlib colormap by combining multiple colormaps.
+
+    Parameters
+    ----------
+        cmaps: list of Colormap or colormap names (str)
+            list of colormaps to combine.
+        nodes: list or numpy array of floats. Defaults: equal divisions.
+            The blending points between colormaps, in the range [0, 1].
+        n_rgb_levels: int. Defaults: 256.
+            Number of RGB levels for each colormap segment.
+        combined_cmap_name: str. Defaults: "combined_cmap".
+            name of the combined Colormap.
+
+    Returns
+    -------
+        Colormap: The composite colormap.
+
+    Raises
+    ------
+        TypeError: If the list contains mixed datatypes or invalid
+            colormap names.
+        ValueError: If the cmaps contain only one single colormap,
+            or if the number of nodes is not one less than the number
+            of colormaps, or if the nodes do not contain incrementing values
+            between 0.0 and 1.0.
+
+    Note
+    ----
+        The colormaps are combined from bottom to top.
+
+    References
+    ----------
+        - https://stackoverflow.com/questions/31051488/combining-two-matplotlib-colormaps/31052741#31052741
+
+    Examples
+    --------
+        Using predefined colormap names::
+        >>> custom_cmap_1 = combine_cmaps(
+            ["ocean", "prism", "coolwarm"], nodes=[0.2, 0.75]
+        )
+
+        Using Colormap objects::
+        >>> cmap_0 = plt.get_cmap("Blues")
+        >>> cmap_1 = plt.get_cmap("Oranges")
+        >>> cmap_2 = plt.get_cmap("Greens")
+        >>> custom_cmap_2 = combine_cmaps([cmap_0, cmap_1, cmap_2])
+
+    """
+    # Check colormap datatype and convert to list[Colormap]
+    if len(cmaps) == 1:
+        raise ValueError("Cannot combine a single colormap.")
+    for cm in cmaps:
+        if not isinstance(cm, (Colormap, str)):
+            raise TypeError(f"Unsupported colormap type: {type(cm)}.")
+    _cmaps: list[Colormap] = [
+        cm if isinstance(cm, Colormap) else mpl.colormaps[cm] for cm in cmaps
+    ]
+
+    # Generate default nodes for equal separation
+    if nodes is None:
+        nodes = np.linspace(0, 1, len(_cmaps) + 1)[1:-1].tolist()
+    elif isinstance(nodes, np.ndarray):
+        nodes = nodes.tolist()
```

Comment:
this branch is detected as unreachable by mypy.
That's because you've declared that the input should be `None` or a `list[float]`. To be clear, the issue isn't the type annotation, it's the code logic; you got the gist of it right: indeed, we should be flexible on what we accept as inputs (here I'm even arguing that we can afford to be *more* flexible than what we declare in type annotations), and then narrow type the actual types of internal variables for simplicity. However, I think it'd be easier to use a `np.ndarray` internally rather than a list: that way, we get support for numpy inputs for free !
This is why in an earlier suggestion I proposed to use a different name for the internal variable so its type isn't bound to that of the argument's annotation. Here's my suggestion again
```suggestion
    nodes_arr = np.array(nodes)
```

Reply:
Yes it makes sense, I also noticed the `mypy` error for Python 3.9 and Python 3.12 here. However the errors doesn't show up locally with `pre-commit` (maybe Python version problem as I was using Python 3.11). I would try your suggested fixes and thanks a ton again.